### PR TITLE
fix(page): fix page.#scrollIntoViewIfNeeded method

### PR DESCRIPTION
**Pull request by abozhilov**
This patch fixes page.#scrollIntoViewIfNeededso that it works with devtools protocol.
Now it blocks the main thread and waits until the scrolling action finishes.

Issues: #8627, #1805

BREAKING CHANGE: page.#scrollIntoViewIfNeeded throws erros which come from the internal implementation.
 - `Protocol error (DOM.scrollIntoViewIfNeeded): Node is detached from document`
 - `Protocol error (DOM.scrollIntoViewIfNeeded): Node does not have a layout object`
Also now it does work with TextNode properly and does not throw an error.

### CODE REVIEW

**Comment by OrKoN** [comment]
it looks like Firefox implementation does not support `DOM.scrollIntoViewIfNeeded` :-/ let's fallback to the script evaluation if `DOM.scrollIntoViewIfNeeded` fails?

**Comment by abozhilov** [comment]
@OrKoN Not sure it's the best approach as it'll introduce differences between chrome/firefox. `DOM.scrollIntoViewIfNeeded` works with text nodes, while the former implementation doesn't. 

See this test case: https://github.com/puppeteer/puppeteer/blob/main/test/src/elementhandle.spec.ts#L206  
Even that the name says it should work, currently it doesn't due to `#scrollIntoViewIfNeeded` implementation. I've fixed this test case with my PR, but obviously it doesn't pass in Firefox. 

I think we should log a feature request in Firefox bug tracker for  `DOM.scrollIntoViewIfNeeded`  and  my PR could wait, unless you have a better solution of the problem.

**Comment by OrKoN** [comment]
I don't think Firefox is going to look into extending the CDP support anytime soon unfortunately. cc @whimboo Henrik, is it something that could be a priority for you?

**Comment by whimboo** [comment]
I doubt that we could take that unless someone is interested to actually help with the implementation. Also [`DOM.scrollIntoViewIfNeeded`](https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-scrollIntoViewIfNeeded) is marked as `experimental` which makes it lesser a priority. Sorry.

**Comment by abozhilov** [comment]
@OrKoN I'll investigate it a bit further and I think I'll come up with a solution which would work in both envs.

**Comment by abozhilov** [comment]
I ended up with this implementation which keeps backward compatibility as much as possible and it falls back to `Element.scrollIntoView` if  `DOM.scrollIntoViewIfNeeded` is not supported. Anyway with the latest release `15.3.1` a lot of tests fail and it's hard for contributing.

```ts
  async #scrollIntoViewIfNeeded(): Promise<void> {
    const error = await this.evaluate(
      async (element): Promise<string | undefined> => {
        if (!element.isConnected) {
          return 'Node is detached from document';
        }
        if (element.nodeType !== Node.ELEMENT_NODE) {
          return 'Node is not of type HTMLElement';
        }
        return;
      }
    );

    if (error) {
      throw new Error(error);
    }

    try {
      await this._client.send('DOM.scrollIntoViewIfNeeded', {
        objectId: this._remoteObject.objectId,
      });
    } catch (_err) {
      // Fallback to Element.scrollIntoView if DOM.scrollIntoViewIfNeeded is not supported
      await this.evaluate(
        async (element, pageJavascriptEnabled): Promise<void> => {
          // force-scroll if page's javascript is disabled.
          if (!pageJavascriptEnabled) {
            element.scrollIntoView({
              block: 'center',
              inline: 'center',
              // @ts-expect-error Chrome still supports behavior: instant but
              // it's not in the spec so TS shouts We don't want to make this
              // breaking change in Puppeteer yet so we'll ignore the line.
              behavior: 'instant',
            });
          }
          const visibleRatio = await new Promise(resolve => {
            const observer = new IntersectionObserver(entries => {
              resolve(entries[0]!.intersectionRatio);
              observer.disconnect();
            });
            observer.observe(element);
          });
          if (visibleRatio !== 1.0) {
            element.scrollIntoView({
              block: 'center',
              inline: 'center',
              // @ts-expect-error Chrome still supports behavior: instant but
              // it's not in the spec so TS shouts We don't want to make this
              // breaking change in Puppeteer yet so we'll ignore the line.
              behavior: 'instant',
            });
          }
        },
        this.#page.isJavaScriptEnabled()
      );
    }
  }
```

**Comment by OrKoN** [comment]
@abozhilov I think 15.3.1 only contains type and documentation changes. The amount of flaky failures (which is unfortunately not zero) should be the same (but we removed automatic retries and we aim at fixing those). Could you update your PR with the latest code version?

**Comment by abozhilov** [comment]
I've just updated my PR.

File: src/common/ElementHandle.ts [hunk 1]
```diff
@@ -240,50 +240,61 @@ export class ElementHandle<
...
+              block: 'center',
+              inline: 'center',
+              // @ts-expect-error Chrome still supports behavior: instant but
+              // it's not in the spec so TS shouts We don't want to make this
+              // breaking change in Puppeteer yet so we'll ignore the line.
+              behavior: 'instant',
+            });
+          }
+          const visibleRatio = await new Promise(resolve => {
+            const observer = new IntersectionObserver(entries => {
```
Comment (OrKoN) [inline src/common/ElementHandle.ts]: can we use IntersectionObserver to detect when the scroll is done?

Comment (abozhilov) [reply src/common/ElementHandle.ts]: I've tried and the race condition is still there. In order to prevent it you need once the node intersects the viewport you'd need to to something like that: 
`setTimeout(() => resolve(true), 100)` which is not great I think.

File: src/common/ElementHandle.ts [hunk 2]
```diff
@@ -240,50 +240,61 @@ export class ElementHandle<
...
+        return;
+      }
     );
 
     if (error) {
       throw new Error(error);
     }
+
+    try {
+      await this._client.send('DOM.scrollIntoViewIfNeeded', {
```
Comment (OrKoN) [inline src/common/ElementHandle.ts:259]: Can we be sure that the scroll is over once this resolves? In the previous version you had a test with an element that is outside of the viewport, should we bring the test back?

File: test/src/click.spec.ts [hunk 3]
```diff
@@ -164,7 +164,10 @@ describe('Page.click', function () {
     await page.goto(server.PREFIX + '/offscreenbuttons.html');
     const messages: any[] = [];
     page.on('console', msg => {
-      return messages.push(msg.text());
+      if (msg.type() === 'log') {
```
Comment (OrKoN) [inline test/src/click.spec.ts:167]: why was this change needed?

File: src/common/ElementHandle.ts [hunk 4]
```diff
@@ -240,50 +240,61 @@ export class ElementHandle<
...
-            // @ts-expect-error Chrome still supports behavior: instant but
-            // it's not in the spec so TS shouts We don't want to make this
-            // breaking change in Puppeteer yet so we'll ignore the line.
-            behavior: 'instant',
-          });
-        }
-        return false;
-      },
-      this.#page.isJavaScriptEnabled()
+        return;
```
Comment (OrKoN) [inline src/common/ElementHandle.ts:250]: nit: return not needed?

File: test/src/click.spec.ts [hunk 5]
```diff
@@ -164,7 +164,10 @@ describe('Page.click', function () {
...
+        return messages.push(msg.text());
+      }
+      return;
```
Comment (OrKoN) [inline test/src/click.spec.ts:170]: nit: return not needed?

File: src/common/ElementHandle.ts (hunk 2 above)
Comment (abozhilov) [reply src/common/ElementHandle.ts:259]: I decided that it's redundant to add an additional test, since there is a test called `'should click offscreen buttons'`. I've improved it a bit in this PR.

File: test/src/click.spec.ts (hunk 3 above)
Comment (abozhilov) [reply test/src/click.spec.ts:167]: Yes, that's because in Firefox when I call `await this._client.send('DOM.scrollIntoViewIfNeeded')` the error appears in the console and the test case should observe only the log messages.

File: src/common/ElementHandle.ts (hunk 4 above)
Comment (abozhilov) [reply src/common/ElementHandle.ts:250]: If I omit it there is a TS error, because not all paths in the code return value.

File: test/src/click.spec.ts (hunk 5 above)
Comment (abozhilov) [reply test/src/click.spec.ts:170]: Same as above.

**Comment by OrKoN** [comment]
Do you think we still need to mark this PR as a breaking change? It looks backward compatible.

**Review by OrKoN — APPROVED** [review]

**Comment by OrKoN** [comment]
Also, it looks like a few elementhandle.spec.ts tests broke due to the changes to the test file

**Comment by abozhilov** [comment]
@OrKoN It's backward compatible, but to be 100% sure I'll have a look at the test cases which you've mentioned.

**Comment by abozhilov** [comment]
@OrKoN I've just reverted `offscreenbuttons.html` to it's previous version, and now the failing tests would work.

**Comment by paliwal81** [comment]
Hi @OrKoN, I would like to know by when we can expect this pull request to be merged. 
Currently my e2e test cases are failing because of the same issue (as UI form is long and elements are visible on view port).
